pkg/vault/vault.go

```diff
@@ -106,14 +114,81 @@ func (v *Vault) ReadSecret(path string) ([]byte, error) {
 
 	data, ok := secret.Data["value"].(string)
 	if !ok {
-		return nil, fmt.Errorf("secret data type is not string, found %T value: %#v", secret.Data["value"], secret.Data["value"])
+		return nil, fmt.Errorf("secret data type is not string, found %T value: %#v at path: %s", secret.Data["value"], secret.Data["value"], path)
 	}
 
 	return []byte(data), nil
 }
 
+func (v *Vault) manageTokenLifecycle(ctx context.Context, authToken *hashivault.Secret) error {
+	authTokenWatcher, err := v.client.NewLifetimeWatcher(&hashivault.LifetimeWatcherInput{
+		Secret: v.token,
+	})
+	if err != nil {
+		return fmt.Errorf("error initializing auth token lifetime watcher: %v", err)
+	}
+
+	go authTokenWatcher.Start()
+	defer authTokenWatcher.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+
+		case <-authTokenWatcher.DoneCh():
+			// Token failed to renew (e.g expired), re-auth required
+			return nil
+
+		case renewalInfo := <-authTokenWatcher.RenewCh():
+			// Token was successfully renewed
+			if renewalInfo.Secret.Auth != nil {
+				level.Debug(v.logger).Log("msg", fmt.Sprintf("token renewed, new lease: %d", renewalInfo.Secret.Auth.LeaseDuration))
+				v.metrics.authLeaseRenewalTotal.Inc()
+			}
+		}
+	}
+}
+
+func (v *Vault) RenewTokenLease(ctx context.Context) error {
+	for ctx.Err() == nil {
+		lfcErr := v.manageTokenLifecycle(ctx, v.token)
+		if lfcErr != nil {
+			level.Error(v.logger).Log("msg", fmt.Sprintf("unable to manage token lifecycle: %v", lfcErr))
+			// We don't want to turn Mimir into an unready state if Vault fails here
+			<-ctx.Done()
```

Comment:
It's strange to do this waiting for context finish here, because `Vault` doesn't implement Service. I'd suggest to do the blocking in `modules.go` instead, in function calling `RenewTokenLease`. Alternatively, we can make `Vault` to implement Service.